satcom/csp/csp_packet.py

```diff
@@ -0,0 +1,188 @@
+        offset += FLEN_ADDR
+
+        val = (hdr << offset) & bitmask32
+        cls.destination_port = val >> (32 - FLEN_PORT)
+        offset += FLEN_PORT
+
+        val = (hdr << offset) & bitmask32
+        cls.source_port = val >> (32 - FLEN_PORT)
+        offset += FLEN_PORT
+
+        # not implemented, so ignored
+        val = (hdr << offset) & bitmask32
+        _ = val >> (32 - FLEN_FLAGS)
+        offset += FLEN_FLAGS
+
+        obj = cls(
+            priority = cls.priority,
+            source = cls.source,
+            destination = cls.destination,
+            destination_port = cls.destination_port,
+            source_port = cls.source_port
+        )
+
+        return obj
+    
+class CSPPacket():
+    def __init__(self, data: bytearray, header=None):
+        self.header = header or CSPPacketHeader()
+        self._data = data
+
+    @property
+    def data(self):
+        """Protects data from being modified after instantiation"""
+        return self._data
+
+    def err(self):
+        """Throws an error if any params are out of bounds"""
+        if self.header.err() is not None:
+            return self.header.err()
+        return None
+
+    def _max_packet_length(self, max_data_size: int):
+        """Returns the maximum possible packet size based on provided max data size"""
+        return HEADER_LENGTH_BYTES + max_data_size
+
+    def _make_buffer(self, max_data_size: int):
+        """Initializes a new byte slice appropriate for a full CSP packet"""
+        return bytearray(self._max_packet_length(max_data_size))
+
+    def to_bytes(self):
+        """Encodes CSP packet and data into bytes"""
+        buf = self._make_buffer(len(self.data))
+        buf[:HEADER_LENGTH_BYTES] = self.header.to_bytes()
+        buf[HEADER_LENGTH_BYTES:] = self.data
+
+        return buf
+
+    @classmethod
+    def from_bytes(cls, bs: bytearray):
+        """Hydrates the CSP packet object from provided byte array"""
+        if len(bs) < HEADER_LENGTH_BYTES:
+            raise ValueError('insufficient data')
+
+        hbs, dbs = bs[0:HEADER_LENGTH_BYTES], bs[HEADER_LENGTH_BYTES:]
+        hdr = CSPPacketHeader.from_bytes(hbs)
+
+        obj = cls(
+            header = hdr,
+            data = dbs
+        )
+
+        return obj
+
+    def write_packet(self):
+        """Writes n bytes to CSP packet"""
+        enc = self.to_bytes()
+        enc_len = len(enc)
+        dst = BytesIO()
+
+        n, err = dst.write(enc)
+
+        if err is not None:
+            return self.err()
+
+        if n != enc_len:
+            return ValueError(f'CSP wrote failed: want {enc_len} bytes, got {n}')
+
+        return None
+
+    def read_packet(self, buf: bytearray):
+        """Reads CSP packet using supplied buffer"""
+        src = BytesIO(buf)
+        n, err = src.getvalue()
+
+        if err is not None:
+            return self.err()
+        buf = buf[:n]
+
+        return self.from_bytes(buf)
```

Comment:
I don't think these methods are necessary. The way write_packet is written, the destination is never actually accessible to the client.